picard/util/pipe.py

```diff
@@ -0,0 +1,360 @@
+    @abstractmethod
+    def PIPE_DIRS(cls):
+        raise NotImplementedError
+
+    def __init__(self, app_name: str, app_version: str, args=None, forced_path=None):
+        if args is None:
+            self._args = tuple()
+        else:
+            try:
+                self._args = tuple(args)
+            except TypeError as exc:
+                raise PipeErrorInvalidArgs(exc) from None
+
+        if not self._args:
+            self._args = (self.MESSAGE_TO_IGNORE,)
+
+        if not isinstance(app_name, str) or not isinstance(app_version, str):
+            raise PipeErrorInvalidAppData
+        elif IS_WIN:
+            app_version = app_version.replace(".", "-")
+
+        if forced_path:
+            self._paths = (forced_path,)
+        elif IS_WIN or os.getenv("HOME"):
+            self._paths = self.__generate_filenames(app_name, app_version)
+            self.path_was_forced = False
+        else:
+            self._paths = (NamedTemporaryFile(delete=False).name,)
+            self.path_was_forced = True
+            log.debug("Pipe path had to be mocked by a temporary file")
+        self.is_pipe_owner: bool = False
+
+        self.__thread_pool = concurrent.futures.ThreadPoolExecutor()
+
+    def _remove_temp_attributes(self):
+        del self._args
+        del self._paths
+
+    def __generate_filenames(self, app_name: str, app_version: str):
+        _pipe_names = []
+
+        for dir in self.PIPE_DIRS:
+            if dir:
+                _pipe_names.append(os.path.join(os.path.expanduser(dir),
+                                                        f"{app_name}_v{app_version}_pipe_file"))
+
+        if _pipe_names:
+            return _pipe_names
+
+        raise PipeErrorNoDestination
+
+    def _reader(self) -> str:
+        raise NotImplementedError()
+
+    def _sender(self, message) -> bool:
+        raise NotImplementedError()
+
+    def read_from_pipe(self, timeout_secs: Optional[float] = None) -> List[str]:
+        if timeout_secs is None:
+            timeout_secs = self.TIMEOUT_SECS
+
+        reader = self.__thread_pool.submit(self._reader)
+        out = []
+
+        try:
+            res = reader.result(timeout=timeout_secs)
+            if res:
+                res = res.split("\n")
+                for r in res:
+                    if res == self.MESSAGE_TO_IGNORE:
+                        out = []
+                        break
+                    elif r:
+                        out.append(r)
+
+        except concurrent.futures._base.TimeoutError:
+            # hacky way to kill the file-opening loop
+            self.send_to_pipe(self.MESSAGE_TO_IGNORE)
+
+        if out:
+            return out
+
+        return [self.NO_RESPONSE_MESSAGE]
+
+    def send_to_pipe(self, message: str, timeout_secs: Optional[float] = None) -> bool:
+        if timeout_secs is None:
+            timeout_secs = self.TIMEOUT_SECS
+
+        # we're sending only filepaths, so it's safe to append newline
```

Comment:
What if a filepath contains `\n`?

Reply:
In my next commit I'll refactor it to `AbstractPipe._MESSAGES_SEPARATOR: str = "\n\n\n\t\t\n"`

Reply:
I don't think that's a clean solution. You see have a separator that may occur in a filepath.

Reply:
I wonder if we could move to a binary protocol, converting "messages" to a sequence of bytes: X bytes for message type (normal or error, future usage), Y bytes for length of payload, Z bytes for payload.

I see multiple advantages:
- no need to escape or separate data (bytes Y)
- it can be extended (bytes X)
- payload can be anything (bytes Z + Y), bytes X can be used to give it a type
- struct.pack() / unpack() on send/receive

As an example, a path could be sent as `(0, 73, b"path_73_bytes_long_path")`, `(255, 0)` could be an error message, `(1, 7, b"command") a specific command.

May be overkill, but it wouldn't require much changes to the current code I think.

For sure, assuming paths (especially on Unix) don't contain "bad" characters, and transmitting those with appended separator doesn't look a good idea to me.

Another approach would be to use `pickle` to (de)serialize objects.

@phw what do you think?

Reply:
I've found out that ['\0' is forbidden in filenames on both Windows and Unix](https://stackoverflow.com/questions/54205087/how-can-i-create-a-file-with-null-bytes-in-the-filename) (`ValueError: embedded null byte`), so we can safely split messages without any need for such deep refactoring. How about that?

Reply:
I think  `\0` works as a separator. As long as we have only paths to transfer this is especially easy. If we start getting different messages we need to introduce some message identifier.

A simple binary format like @zas proposed would work. Or if we really just expect text to be transferred the null termination maybe still works, so only `(0, b"path_73_bytes_long_path", b"\0")`.

Do we expect to need the type? If yes maybe introduce it already but only define a single type path = 0